### PR TITLE
[flang] Add MALLOC and FREE intrinsics for Cray pointers

### PR DESCRIPTION
**Pull request by DavidTruby**
MALLOC and FREE are extensions provided by gfortran, Intel Fortran and
classic flang to allocate memory for Cray pointers. These are used in
some legacy codes such as libexodus.

All the above compilers accept using MALLOC and FREE with integers as
well, despite that this will often signify a bug in user code. We should
accept the same as the other compilers for compatibility.

### CODE REVIEW

**Comment by DavidTruby** [comment]
I'd have liked to add a warning for if the result of MALLOC is assigned to a non-Cray pointer variable, but I wasn't sure quite where to put that test. I assume it can't go next to the free one as you don't know what the result variable is in check-call.cpp. If anyone has any pointers on how to do this, please let me know, thanks!

File: flang/docs/Intrinsics.md [hunk 1]
```diff
@@ -700,7 +700,7 @@ IBCHNG, ISHA, ISHC, ISHL, IXOR
 IARG, IARGC, NARGS, NUMARG
 BADDRESS, IADDR
 CACHESIZE, EOF, FP_CLASS, INT_PTR_KIND, ISNAN, LOC
-MALLOC
+MALLOC, FREE
```
Comment (DavidTruby) [inline flang/docs/Intrinsics.md:703]: Note: the docs say that MALLOC is an extension we already support, but it isn't actually supported at the moment.

**Review by pawosm-arm — APPROVED** [review]
Nice one

File: flang/lib/Semantics/check-call.cpp [hunk 2]
```diff
@@ -1600,6 +1600,18 @@ static void CheckMaxMin(const characteristics::Procedure &proc,
   }
 }
 
+static void CheckFree(evaluate::ActualArguments &arguments,
```
Comment (tblah) [inline flang/lib/Semantics/check-call.cpp:1603]: Please could you add a similar check for the argument of MALLOC. I'm worried that we could have a compiler crash if somebody passes a type that cannot be converted to the runtime function argument type.

**Review by tblah — COMMENTED** [review]
Overall looks good. One small comment

File: flang/lib/Semantics/check-call.cpp (hunk 2 above)
Comment (DavidTruby) [reply flang/lib/Semantics/check-call.cpp:1603]: I think we need this for basically all the intrinsics in the big type table in intrinsics.cpp. I reckon we should do that generally and generically for all of them, as a separate patch to this.

**Review by tblah — APPROVED** [review]

**Pull request merged**